File: api/controllers/comment.go
package controllers

import (
	"encoding/json"
	"net/http"
	"time"

	"github.com/dimfeld/httptreemux"
	"github.com/godiscourse/godiscourse/api/middleware"
	"github.com/godiscourse/godiscourse/api/models"
	"github.com/godiscourse/godiscourse/api/session"
	"github.com/godiscourse/godiscourse/api/views"
)

type commentRequest struct {
	TopicID string `json:"topic_id"`
	Body    string `json:"body"`
}

type commentImpl struct{}

func registerComment(router *httptreemux.TreeMux) {
	impl := &commentImpl{}

	router.POST("/comments", impl.create)
	router.POST("/comments/:id", impl.update)
	router.POST("/comments/:id/delete", impl.destory)
	router.GET("/topics/:id/comments", impl.comments)
}

func (impl *commentImpl) create(w http.ResponseWriter, r *http.Request, _ map[string]string) {
	var body commentRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		views.RenderErrorResponse(w, r, session.BadRequestError(r.Context()))
		return
	}
	if comment, err := middleware.CurrentUser(r).CreateComment(r.Context(), body.TopicID, body.Body); err != nil {
		views.RenderErrorResponse(w, r, err)
	} else {
		views.RenderComment(w, r, comment)
	}
}

func (impl *commentImpl) update(w http.ResponseWriter, r *http.Request, params map[string]string) {
	var body commentRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		views.RenderErrorResponse(w, r, session.BadRequestError(r.Context()))
		return
	}
	if comment, err := middleware.CurrentUser(r).CreateComment(r.Context(), params["id"], body.Body); err != nil {
		views.RenderErrorResponse(w, r, err)
	} else {
		views.RenderComment(w, r, comment)
	}
}

func (impl *commentImpl) destory(w http.ResponseWriter, r *http.Request, params map[string]string) {
	if err := middleware.CurrentUser(r).DeleteComment(r.Context(), params["id"]); err != nil {
		views.RenderErrorResponse(w, r, err)
	} else {
		views.RenderBlankResponse(w, r)
	}
}

func (impl *commentImpl) comments(w http.ResponseWriter, r *http.Request, params map[string]string) {
	offset, _ := time.Parse(time.RFC3339Nano, r.URL.Query().Get("offset"))
	if topic, err := models.ReadTopic(r.Context(), params["id"]); err != nil {
		views.RenderErrorResponse(w, r, err)
	} else if topic == nil {
		views.RenderErrorResponse(w, r, session.NotFoundError(r.Context()))
	} else if comments, err := topic.ReadComments(r.Context(), offset); err != nil {
		views.RenderErrorResponse(w, r, err)
	} else {
		views.RenderComments(w, r, comments)
	}
}

File: web/src/api/comment.js
class Comment {
  constructor(api) {
    this.api = api;
  }

  index(id) {
    return this.api.axios.get(`/topics/${id}/comments`).then((resp) => {
      return resp.data;
    });
  }

  create(params) {
    const data = {topic_id: params.topic_id, body: params.body};
    return this.api.axios.post('/comments', data).then((resp) => {
      return resp.data;
    });
  }

  update(id, params) {
    const data = {topic_id: params.topic_id, body: params.body};
    return this.api.axios.post(`/comments/${id}`, data).then((resp) => {
      return resp.data;
    });
  }

  show(id) {
    return this.api.axios.get(`/comments/${id}`).then((resp) => {
      return resp.data;
    });
  }

  delete(id) {
    return this.api.axios.delete(`/comments/${id}`).then((resp) => {
      return resp;
    });
  }
}

export default Comment;
